internal/cloudapi/v2/handler.go

```diff
@@ -888,6 +891,167 @@ func (h *apiHandlers) getComposeManifestsImpl(ctx echo.Context, id string) error
 	return ctx.JSON(http.StatusOK, resp)
 }
 
+// sbomsFromOSBuildJob extracts SBOM documents from dependencies of an OSBuild job.
+func sbomsFromOSBuildJob(w *worker.Server, osbuildJobUUID uuid.UUID) ([]ImageSBOM, error) {
+	var osbuildJobResult worker.OSBuildJobResult
+	osbuildJobInfo, err := w.OSBuildJobInfo(osbuildJobUUID, &osbuildJobResult)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get results for OSBuild job %q: %v", osbuildJobUUID, err)
+	}
+
+	pipelineNameToPurpose := func(pipelineName string) (ImageSBOMPipelinePurpose, error) {
+		if slices.Contains(osbuildJobResult.PipelineNames.Payload, pipelineName) {
+			return ImageSBOMPipelinePurposeImage, nil
+		}
+		if slices.Contains(osbuildJobResult.PipelineNames.Build, pipelineName) {
+			return ImageSBOMPipelinePurposeBuildroot, nil
+		}
+		return "", fmt.Errorf("Pipeline %q is not listed as either a payload or build pipeline", pipelineName)
+	}
+
+	// SBOMs are attached to the depsolve job results.
+	// Depsolve jobs are dependencies of Manifest job.
+	// Manifest job is a dependency of OSBuild job.
+	manifesJobInfo, _, err := manifestJobResultsFromJobDeps(w, osbuildJobInfo.Deps)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get manifest job info for OSBuild job %q: %v", osbuildJobUUID, err)
+	}
+
+	var imageSBOMs []ImageSBOM
+	for _, manifestDepUUID := range manifesJobInfo.Deps {
+		depJobType, err := w.JobType(manifestDepUUID)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get job type for dependency %q: %v", manifestDepUUID, err)
+		}
+
+		if depJobType == worker.JobTypeDepsolve {
+			var depsolveJobResult worker.DepsolveJobResult
+			_, err = w.DepsolveJobInfo(manifestDepUUID, &depsolveJobResult)
+			if err != nil {
+				return nil, fmt.Errorf("Failed to get results for depsolve job %q: %v", manifestDepUUID, err)
+			}
+
+			if depsolveJobResult.SbomDocs == nil {
+				return nil, fmt.Errorf("depsolve job %q: missing SBOMs", manifestDepUUID)
+			}
+
+			for pipelineName, sbomDoc := range depsolveJobResult.SbomDocs {
+				purpose, err := pipelineNameToPurpose(pipelineName)
+				if err != nil {
+					return nil, fmt.Errorf("Failed to determine purpose for pipeline %q: %v", pipelineName, err)
+				}
+
+				var sbomType ImageSBOMSbomType
+				switch sbomDoc.DocType {
+				case sbom.StandardTypeSpdx:
+					sbomType = ImageSBOMSbomTypeSpdx
+				default:
+					return nil, fmt.Errorf("Unknown SBOM type %q attached to depsolve job %q", sbomDoc.DocType, manifestDepUUID)
+				}
+
+				imageSBOMs = append(imageSBOMs, ImageSBOM{
+					PipelineName:    pipelineName,
+					PipelinePurpose: purpose,
+					Sbom:            sbomDoc.Document,
+					SbomType:        sbomType,
+				})
+			}
+
+			// There should be only one depsolve job per OSBuild job
+			break
```

Comment:
Should this `break` be inside the `depJobType == worker.JobTypeDepsolve` block?  The way it is now, only one the first `manifestJobInfo.Deps` element is considered.  Is that always a depsolve job and nothing else?  I guess no, otherwise we'd just be using `manifestJobInfo.Deps[0]` right?

Reply:
> Should this break be inside the depJobType == worker.JobTypeDepsolve block?
It actually is inside block as far as I can tell :thinking: 

So the break is at the correct place, but this is a bit of a premature optimization based on the current implementation. Let me explain.

I implemented `sbomsFromOSBuildJob()` in a way, so that it uses the same approach to get SBOMs from `OSBuildJob`, regardless if it is from a Koji or non-Koji compose.

Just for the completeness, for Koji composes, the `DepsolveJob` is a dependency of the `OSBuildJob`, so that it can access SBOMs and upload them to Koji. For non-Koji composes, the `DepsolveJob` is not a direct dependency of the `OSBuildJob` and one must go through the `ManifestJob` as to get to `DepsolveJob`.

There's always only a single `DepsolveJob` in the compose job chain, thus the break. But there may be more than one item in `manifestJobInfo.Deps` and `DepsolveJob` does not have a fixed index.

Reply:
Right.  Thanks for the explanation.  I think I just read it wrong :sweat_smile: